Fixing widgets positions

imenattatra opened the pull request:
## What type of PR is this? 

- [x] Bug Fix


## Description
This PR is a fix following the feature about showing widgets according to parameter value. 

### The problem

When widgets were disappearing and then appearing again, it occurred that they re-appeared in a wrong position far from their purpose , for example a widget about water meters appearing with the widgets about heat meters.

### The tested solutions

Since we don't have the concept of sections in redash (we could have had different sections using the same used package but this is not the used also ) and since also we don't controle directly the code that re-ajusts the layouts since it is done by a third party package `react-grid-layout`, there was no easy solution to deal with that, i tried many options that seemed in the beginning lighter that the solution that we have now:

- hide widgets using css .. this by sending all widgets with a flag that we pass later by to display as hidden or as block
- disable react-grid-layout compaction by setting it as null and try to handle compaction by our code - [related commit to ignore ](https://github.com/metr-systems/redash/commit/accd629cd0c904a343f1c37bfa5184c15f3f9ef6) -
- compare against initial set of groups of widgets - [related commit to ignore](https://github.com/metr-systems/redash/commit/390f98a83096ad8e791385d7fd4e4e6a262a6cfb) - 

### The final solution

We ended up taking a more drastic solution which is we fix the layout according to the order of all the widgets.

We have the complete order of all widgets going from **left to right to bottom** and we have the list of layouts to show, we also know that every layout represents one widget and has its same id. 

So we go through the ordered widget ids list and for each of then we get the exact layout and fix how it appears on the UI, either by moving it to left or up or left-up and make sure it appears near its group of widgets. The function resposible of that is called `keepLayoutsOrder` and I invite you to check the related tests for it bellow because in the tests comments you find visual documentation for how it is changing layouts positions and its limitations.


## How is this tested?

- [x] Unit tests (jest)

I may be adding more tests to confirm that UI close to our most important dashboards are handled well.

## Note

This code is deployed on staging :)

imenattatra committed:
[WIP][JFD]

committing this just for documenting this attempt to resolve the widget positions issue.

The logic tried here, is to disable compacting the layouts by the third party package and then compacting them with our function ´compactLayoutsHorizontally´but this is not working because the layouts values received when in ´onLayoutChange´ are not correct ... need to find a reset solution maybe

imenattatra committed:
Revert "[WIP][JFD]"

This reverts commit accd629cd0c904a343f1c37bfa5184c15f3f9ef6.

imenattatra committed:
[[WIP]

When the layouts changes according to parameter value, some of them need to disappear and then appear again, they appear in the wrong place.
In this code we handle only the one and only known senario so far for this case, the solution here a drafty one and it would be better if we find better solutions in the futur

imenattatra committed:
make sure our layout keep same positioning order

whenever they disappear and appear when we change the parameter value

imenattatra committed:
update function that correct widgets positionning when hidden/reappearing

imenattatra committed:
add more cases of when widgets have inequal height

imenattatra committed:
optimize  function

imenattatra committed:
remove console.log

imenattatra committed:
case when we have many widgets per line

imenattatra committed:
add integration senario for reorganizing widgets making sure the order is maintained

imenattatra committed:
update ´keepLayoutsOrder´for better code

imenattatra committed:
update ´keepLayoutsOrder´by extracting some code for a better readability

charinass commented:
Oh wow, i really enjoyed reading this PR description. It's describing to me as like a child. I love it! I don't know much about ".js" and ".jsx" but by the looks of it, it seems you handled it well! Love, love, love the visaulization! Thank you! 🚀

thiagogds reviewed (approved):
This PR has the most fantastic test I've ever seen! I loved them so much. 

It's tough to understand the code without the context of the visualizations and layouts. But the tests make it easy to understand and create the trust that everything works.

Also, the tests are super simple to create, and the visualization in the comments is the best.

Also, thanks for documenting your tries and why you decided on this solution

Thanks for the fantastic PR.

berinhard commented on client/app/components/dashboards/utils.js:
```diff
@@ -0,0 +1,61 @@
+    usedColumns += layout.w;
+  };
+
+  const shouldMoveToNewLine = (isWidgetTextbox, isPreviousTextbox, previousLayout, layout) =>
+    isWidgetTextbox ||
+    isPreviousTextbox ||
+    previousLayout.w + layout.w > cfg.columns ||
+    usedColumns + layout.w > cfg.columns;
```
The decision of defining anonymous functions to increase the meaning of what is really happening is awesome!

berinhard commented on client/app/components/dashboards/utils.js:
```diff
@@ -0,0 +1,61 @@
+      const previousWidget = getWidgetById(previousLayout.i);
+      const isPreviousTextbox = isTextbox(previousWidget);
+
+      if (shouldMoveToNewLine(isWidgetTextbox, isPreviousTextbox, previousLayout, layout)) {
+        moveToNewLine(layout);
+      } else {
+        placeOnSameLine(layout, previousLayout);
+      }
```
This is so readable!!! Amazing, for real

berinhard commented on client/app/components/dashboards/utils.test.js:
```diff
@@ -0,0 +1,1015 @@
+
+describe("keepLayoutsOrder", () => {
+  it("widget moves to left when there is space", () => {
+    /** 
+        full layout looks like this:
+        ------------- -------------
+        |           | |           |
+        | vis1      | |  vis2     |
+        ------------- -------------
+        ------------- -------------
+        |           | |           |
+        | vis3      | |  vis4     |
+        ------------- -------------
+
+        layout with hidden widget 3 look like this:
+
+        ------------- -------------
+        |           | |           |
+        | vis1      | |  vis2     |
+        ------------- -------------
+        -------------
+        |           | 
+        | vis4      | 
+        ------------- 
+        */
```
I simply LOVED the having this visualization as the "docstring" of the test. This was very thoughtful and will enable other people to maintain these tests if needed too.

berinhard reviewed (approved):
This PR is a beautiful piece of work! The PR description is great and the code mirrors the care and attention you're putting to this work. Kudos @imenattatra!

My only complaint is: I missed a gif (just kidding)

![kudos](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExaHV5cmhlZ25tZzA2MWw5dGVwMWxxMHJrb3dhODFwN3k1dmo3eHh5ciZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/3ohc16GbYSmfF4HvB6/giphy-downsized.gif)

GulGulHub reviewed (approved):
I can only agree with everything said above: freaking amazing PR!!! What a great documentation of everything! Thank you so much for the details and description and visualization!

imenattatra committed:
allow to change widgets positionning while editing

imenattatra committed:
use smaller widgets representation in docstring

imenattatra commented:
Thank you all for your kind and awesome feedbacks 😃  🫶  I am so happy 😃 ❤️ 

I want also to inform you that I added i[n this commit](https://github.com/metr-systems/redash/pull/9/commits/00c2c1813e5a5f7b25a003df52f5a0178398aece)  a condition to allow people to change the order of widgets while editing because otherwise they cannot ...and since they show all widgets while editing, there should be no issue (even if we are discussing the possibility to enforce showing all widgets while editing if needed)

And this is the GIF you are missing @berinhard 

![](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExdGlqaHJraXN5ZGF2amMxazRpOTAwN2ZlMjBhYTF1ajBld200M3FmcCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/l0amJzVHIAfl7jMDos/giphy.gif)

imenattatra committed:
update comments

imenattatra committed:
save layouts order after editing

We update the function ´setEditing´ that is executed after clicking on ´Done Editing´ button to set editing
to false.
We update it to savethe edited layout order, because our UI should be using that new order after  edition.
We pass ´editedlayoutsOrder´ and ´setEditedlayoutsOrder´ to DashboardPage via the custom hook ´useDashboard´
and there we can pass them to´DashboardGrid´ where we need to use them

imenattatra committed:
update ´onLayoutChange´ to use the edited layout whenever needed

imenattatra committed:
update naming and make code lighter in ´onLayoutChange´ in dashboard grid

imenattatra committed:
extract the logic of applying the layouts order when not editing

imenattatra committed:
move ´calculateLayoutsOrder´ to utils and add tests for it

imenattatra commented:
There was actually further issues with the edition 😄 and we had to fix it by the simple procedure which can be explained by the following :
- enable people to change layouts ordering while on edit mode (already done earlier)
- save the layouts order after clicking on "Done Editing"... which we can call "edited layouts order", [related commit](https://github.com/metr-systems/redash/pull/9/commits/b651eeb431ebfebd67cf3619a4344f831dee25f1)
- in the dashboard grid, initialise the layouts order to be the default one that we get from saved_all_widgets, but then make sure to update it with "edited layouts order" after an edit ... and reset "edited layouts order" so that we can use it again in next edit, related commits: https://github.com/metr-systems/redash/pull/9/commits/19705f6c12333556ad3dd3d65429924e09451d66, https://github.com/metr-systems/redash/pull/9/commits/967bd16e01f7a253124023519e63c5c3ef023b74, https://github.com/metr-systems/redash/pull/9/commits/689a0080579e1aaa6c5be7b23b7ce7b0c507cffd
- [UPDATE] make sure we also save the new layouts order when clicking on "Done Editing" while the UI is still saving .... it shows "saving ..." and then "saved", [related commit ](https://github.com/metr-systems/redash/pull/9/commits/a5dd310e2d54bb25144f628cb1d87ecec80ff62e)

sorry for changing after it has been approved but i will ask for your review again after few more manual tests and confirmation :D

imenattatra committed:
save edited layout order also when clicking on  while it is saving.

jhlund reviewed (approved):
Wow! :astonished:

thiagogds commented on client/app/pages/dashboards/hooks/utils.test.js:
```diff
@@ -194,3 +195,56 @@ describe("filterWidgets", () => {
+
+describe("calculateLayoutsOrder", () => {
+  test("should return layouts ordered by row and col", () => {
+    const layouts = {
+      "10": { col: 0, row: 0, sizeX: 3, sizeY: 8 },
+      "11": { col: 0, row: 8, sizeX: 3, sizeY: 8 },
+      "12": { col: 0, row: 16, sizeX: 3, sizeY: 4 },
+    };
```
Please change the order to make sure that the tests are working. Now your code could just return the keys of the layouts and it would work.

imenattatra replied:
It was working even after i changed the tests and added a new one with more lines, but I updated the code in [this commit](https://github.com/metr-systems/redash/pull/9/commits/8f1b428598fbaed0d50a808f1648e4e8c60962d1) to be more readable because i also did not feel comfortable with the previous code. Thank you for pointing it out

thiagogds commented on client/app/pages/dashboards/hooks/utils.test.js:
```diff
@@ -194,3 +195,56 @@ describe("filterWidgets", () => {
+  });
+
+  test("should handle layouts with same col but different row", () => {
+    const layouts = {
+      "10": { col: 0, row: 0, sizeX: 3, sizeY: 8 },
+      "11": { col: 0, row: 1, sizeX: 3, sizeY: 8 },
+      "12": { col: 0, row: 2, sizeX: 3, sizeY: 4 },
+    };
```
Same thing about the order

imenattatra replied:
same as [previous comment ](https://github.com/metr-systems/redash/pull/9#discussion_r1818600269)

thiagogds commented on client/app/pages/dashboards/hooks/utils.test.js:
```diff
@@ -194,3 +195,56 @@ describe("filterWidgets", () => {
+  });
+
+  test("should handle layouts with mixed rows and cols", () => {
+    const layouts = {
+      "10": { col: 0, row: 0, sizeX: 3, sizeY: 8 },
+      "11": { col: 2, row: 1, sizeX: 3, sizeY: 8 },
+      "12": { col: 1, row: 2, sizeX: 3, sizeY: 4 },
+    };
```
Also the order

imenattatra replied:
same as [previous comment ](https://github.com/metr-systems/redash/pull/9#discussion_r1818600269)

thiagogds reviewed (changes requested):
Looks good! I did leave a comment in the tests that need some tweaking.

Next time, instead of sticking with the same PR, you could merge this into an epic branch and start a new one just for the editing part. It’ll make things way easier to review.

It might feel a bit off merging like this, but the whole idea behind an epic is that you can break it down into smaller, simpler topics.

imenattatra committed:
use a more readable code for ordering the layouts

imenattatra commented:
> Looks good! I did leave a comment in the tests that need some tweaking.
> 
> Next time, instead of sticking with the same PR, you could merge this into an epic branch and start a new one just for the editing part. It’ll make things way easier to review.
> 
> It might feel a bit off merging like this, but the whole idea behind an epic is that you can break it down into smaller, simpler topics.

Sorry, you're totally right about the epic usage , I thought it was a small change but then things escalated, will remember next time :D , thanks
